### PR TITLE
Changes to class-wc-auth that allows it to work with sso auth

### PR DESCRIPTION
**Pull request by maxdupenois**
### All Submissions:

* [ x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [ x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)? _Except for the use of the `global` for which I couldn't see a way around_
* [ x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This is an initial attempt to fix an issue with the wc-auth endpoint which doesn't work alongside WordPress' SSO implementation. The issue came to light when someone was testing the connection flow for the Facebook for Woocommerce plugin on their site which makes use of the wc-auth endpoint:

https://user-images.githubusercontent.com/1573005/119368147-98443080-bcaa-11eb-9ced-1257233970a7.mov

The password they entered is in fact correct and if they logged in via the normal route then restarted the flow they'd be able to continue. After a bit of digging I realised that the issue was that the user doesn't actually login to their site, when they go to /wp-admin it redirects them to a wordpress.com domain to login. Once they complete their login there they're redirected back to their actual site with some kind of authenticated token.

Wordpress SSO is handled via the Jetpack plugin so this PR aims to ensure that the login hook that Jetpack uses is correctly triggered by the wc-auth endpoint. The below video shows it working:

https://user-images.githubusercontent.com/1573005/119368862-70a19800-bcab-11eb-97df-faee091a4b00.mov

As can be seen it correctly redirects to the wordpress log in then redirects back to the OAuth flow that had originally been started with the logged in user.

> **NB**
> I do not know all the ways in which this endpoint is expected to be used so haven't been able to comprehensively test.

#### Manual Test
1. You'll need a site using the SSO method supplied by Jetpack
2. Visit https:/[YOUR SITE]/wc-auth/v1/authorize/?app_name=WooCommerce%20Integration&user_id=[USER ID]&return_url=[somewhere.com]&callback_url=[somewhereelse.com]&scope=read_write
3. Check that the redirect works

### Other information:

* [ x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ 0] Have you written new tests for your changes, as applicable? - **NB:** I'd like to but there are no existing tests for wc-auth (or at least none that I found with grep for wc-auth or WC_Auth. Is there any specific way you'd like these to be added?
* [ x] Have you successfully run tests with your changes locally?

### Changelog entry

Fixed OAuth flows that use the wc-auth endpoint on sites that use SSO with a redirect for authentication.

### CODE REVIEW

**Comment by maxdupenois** [comment]
With respect to the codesniff about the `global`, I'm not convinced there's a way around it given the way the jetpack code works but I'm no wordpress expert so absolutely willing to change it if there are any suggestions? @roykho?

**Comment by maxdupenois** [comment]
(also the other codesniffs are for code I haven't changed, I am happy to remove the multilines if required but thought it best if you could see the actual relevant change for the review process.)

**Review by vedanshujain — CHANGES_REQUESTED** [review]
Thanks for the PR, I tested and it works but I am not sure about merging this yet, see below:

Login form for WooCommerce should work, irrespective of whether there is SSO enabled or not. Since this works on non JP sites, looks like the issue is because is related to users. I think we should figure out:
1. How can we support JP's SSO properly? Overwriting `$action` shouldn't be the only way.
2. JP should allow WC form to work. This not working indicates that something else is going as well.

I will reach out internally to see what's going on and how can we fix as well. But maybe opening similar issue on JP repo should be our next step.

**Comment by maxdupenois** [comment]
> Thanks for the PR, I tested and it works but I am not sure about merging this yet, see below:
> 
> Login form for WooCommerce should work, irrespective of whether there is SSO enabled or not. Since this works on non JP sites, looks like the issue is because is related to users. I think we should figure out:
> 
> 1. How can we support JP's SSO properly? Overwriting `$action` shouldn't be the only way.
> 2. JP should allow WC form to work. This not working indicates that something else is going as well.
> 
> I will reach out internally to see what's going on and how can we fix as well. But maybe opening similar issue on JP repo should be our next step.

Hi @vedanshujain, thank you for the reply and sorry for the delay in getting back to you. We can try opening a JP repo issue but I'm not sure that it's a JP problem. Unless I've misunderstood (which is certainly possible) the way the SSO works means that the user doesn't have their password stored on their domain, the redirect sends the user to wordpress where they login with their credentials and they're then redirected back to their own domain with an authentication token. As a result a login form that just checks the database on the local domain will never be able to authenticate correctly.

Agree 100% that setting the $action global is decidedly dodgy, can you point me to where the sanitisation takes place in the code so that I can properly understand the flow?

Cheers

**Comment by rawdreeg** [comment]
I am picking this up. I will look at this and open an issue in JP with properly supporting SSO without the need to override $action.

File: includes/class-wc-auth.php [hunk 1]
```diff
@@ -326,6 +325,36 @@ protected function auth_endpoint( $route ) {
...
+				* @since 5.3.0
+				*/
+				if ( defined( 'JETPACK__PLUGIN_DIR' ) ) {
+					require_once JETPACK__PLUGIN_DIR . 'modules/sso.php';
+					// Jetpack SSO uses the global action to determine whether or not
+					// it should insert itself, unfortunately we don't have an action
+					// at this point (likely because it's been sanitised out) so
+					// we need to force set it.
+					global $action;
+					$action = 'login';
```
[the remark this answers is not in the document]
Comment (rawdreeg) [reply includes/class-wc-auth.php:345]: > Setting this is not a good idea at all, if it's sanitized out here and we need it then we should change the sanitization code to allow it back.


@vedanshujain I have had a look at this and for what we're trying to achieve,  we should be able to build the SSO URL with the appropriate action and perform the redirect like so:

```
// Check if the user is using the WordPress.com SSO.
if ( Jetpack::is_module_active( 'sso' ) ) {

  $redirect_url    = $this->build_url( $data, 'authorize' );
  
  // Build the SSO URL.
  $login_url = Jetpack_SSO::get_instance()->build_sso_button_url( array(
	  'redirect_to' =>  rawurlencode ( esc_url_raw ($redirect_url ) ),
	  'action'      => 'login',
  ) );
  
  // Perform the redirect.
  wp_safe_redirect( $login_url ); exit;
}
```

This allows us to complete the flow. I will be preparing a new PR for this.

cc: @budzanowski

**Comment by vedanshujain** [comment]
Closing in favor of #32335

**Pull request closed without merging**